### PR TITLE
feat: Implement Capability Verification System

### PR DESCRIPTION
**Pull request by teslashibe**
# PR: Implement Capability Verification System

## Overview

This PR implements a comprehensive capability verification system that solves the critical problem where workers advertise capabilities they don't actually possess, leading to job failures and network instability.

### Core Infrastructure
- **Health Tracking System** (`internal/capabilities/health/`)
  - Thread-safe capability status tracking
  - Background reconciliation loop with exponential backoff
  - Real-time health status monitoring

- **Verification Engine** (`internal/capabilities/verifier.go`)
  - Panic recovery protection (prevents verifier crashes)
  - 30-second timeout protection per capability
  - Robust error handling and logging

- **Capability Detection** (`internal/capabilities/detector.go`)
  - Configuration-based capability discovery
  - Verifier registration and management
  - Integration with health tracking

### Concrete Verifiers
- **WebScraperVerifier**: HTTP connectivity test to example.com
- **TikTokVerifier**: API transcription test with known video
- **TwitterVerifier**: Search functionality test (supports both credentials and API keys)
- **LinkedInVerifier**: Profile fetch test with Bill Gates profile

### System Integration
- **Dynamic Telemetry**: `/telemetry` endpoint reflects real-time capability health
- **JobServer Integration**: Health tracker passed through initialization chain
- **Startup Integration**: Capability verification runs at worker startup
- **Background Reconciliation**: Unhealthy capabilities automatically retested

### Bug Fixes
- **LinkedIn Search Bug**: Removed incorrect `searchbyquery` capability from LinkedIn detection
- **Twitter API Key Support**: Extended Twitter verifier to handle API key authentication
- **Unverifiable Capabilities**: Capabilities without verifiers now marked as unhealthy
- **Import Cycle Resolution**: Moved shared types to `api/types/` to resolve circular dependencies

### Startup Flow
1. Worker starts → `DetectCapabilities()` called
2. Capabilities detected from configuration
3. Each capability verified with timeout/panic protection
4. Only healthy capabilities advertised to network
5. Background reconciliation loop started

### Runtime Flow
1. Jobs execute using healthy capabilities
2. Job failures update capability health status
3. Telemetry endpoint shows real-time healthy capabilities
4. Background reconciliation retests unhealthy capabilities
5. Auto-recovery when services come back online

### Error Handling
- **Verifier Panics**: Caught and logged, capability marked unhealthy
- **Timeouts**: 30-second limit prevents blocking, capability marked unhealthy
- **Network Issues**: Temporary failures trigger reconciliation attempts
- **Invalid Credentials**: Permanent failures until configuration changes

### 1. Missing Verifiers
- [ ] **Telemetry Verifier**: Currently no verifier for `telemetry` capability
  - **Impact**: Telemetry capability marked as unhealthy by default
  - **Solution**: Create simple verifier that checks internal telemetry functionality
  - **File**: `internal/capabilities/verifiers/telemetry.go`

### 2. Test System Updates
- [ ] **Detector Tests**: Update tests to expect realistic verification outcomes
  - **Impact**: Tests fail because they expect fake credentials to work
  - **Current**: Tests expect all capabilities to be healthy
  - **Reality**: Capabilities with invalid credentials are filtered out
  - **File**: `internal/capabilities/detector_test.go`

- [ ] **Integration Tests**: Add tests for edge cases
  - **Panic Recovery**: Test verifier panic handling
  - **Timeout Protection**: Test slow verifier scenarios
  - **Network Failures**: Test connectivity issues

### 3. LinkedIn Credential Parsing
- [ ] **Robust Parsing**: Improve LinkedIn credential parsing logic
  - **Issue**: Current parsing expects specific map format
  - **Problem**: Test with `[]interface{}{"cred1"}` doesn't parse correctly
  - **Solution**: Add better validation and error handling
  - **File**: `internal/capabilities/detector.go` lines 73-83

### 4. Twitter API Key Verification
- [ ] **Real API Key Testing**: Twitter API key verification is placeholder
  - **Issue**: Current implementation doesn't actually test API keys
  - **Problem**: twitter-scraper library doesn't support API keys directly
  - **Solution**: Implement proper API key validation or use different library
  - **File**: `internal/capabilities/verifiers/twitter.go` lines 140-170

### 5. Performance Improvements
- [ ] **Concurrent Verification**: Run verifications in parallel
  - **Current**: Sequential verification (30s per capability)
  - **Target**: Parallel verification (30s total maximum)
  - **Benefit**: Faster startup times

- [ ] **Configurable Timeouts**: Allow per-capability timeout configuration
  - **Current**: Hard-coded 30-second timeout
  - **Target**: Configurable timeouts in job configuration
  - **Benefit**: Optimize for different capability types

### 6. Monitoring & Observability
- [ ] **Prometheus Metrics**: Add capability health metrics
  - **Metrics**: Capability status, verification duration, failure counts
  - **Integration**: Expose via `/metrics` endpoint
  - **Benefit**: Better operational visibility

- [ ] **Health Dashboard**: Visual capability health monitoring
  - **Features**: Real-time status, historical trends, error details
  - **Integration**: Web UI or API endpoints
  - **Benefit**: Easier debugging and monitoring

### 7. Advanced Features
- [ ] **Graceful Degradation**: Partial capability functionality
  - **Example**: Twitter search works but profile fetch fails
  - **Current**: All-or-nothing capability health
  - **Target**: Granular capability sub-features

- [ ] **Capability Dependencies**: Model capability relationships
  - **Example**: LinkedIn profile depends on LinkedIn auth
  - **Current**: Independent capability verification
  - **Target**: Dependency-aware verification

### 8. Security Enhancements
- [ ] **Credential Rotation**: Support for rotating credentials
  - **Current**: Static credentials loaded at startup
  - **Target**: Dynamic credential updates without restart
  - **Benefit**: Better security practices

- [ ] **Rate Limiting**: Add rate limiting for verification requests
  - **Current**: Unlimited verification requests
  - **Target**: Respect API rate limits during verification
  - **Benefit**: Avoid getting blocked by external services

### Unit Tests
- [x] Basic capability detection logic
- [x] Health tracker functionality
- [x] Verifier system with mocks
- [ ] Edge case handling (panics, timeouts, network failures)

### Integration Tests
- [ ] End-to-end capability verification flow
- [ ] Network failure simulation
- [ ] Credential validation scenarios
- [ ] Background reconciliation testing

### Performance Tests
- [ ] Startup time with multiple capabilities
- [ ] Memory usage under load
- [ ] Concurrent verification performance

### For Existing Deployments
1. **Backward Compatibility**: All changes are backward compatible
2. **Gradual Rollout**: Workers will gradually adopt verification
3. **Monitoring**: Watch for capabilities being filtered out
4. **Credential Validation**: Ensure all credentials are valid

### Configuration Changes
- **No Breaking Changes**: Existing configurations continue to work
- **New Behavior**: Invalid credentials now filtered out automatically
- **Monitoring**: Check `/telemetry` endpoint for capability changes

### High Risk
- **Startup Time**: Verification adds up to 30s per capability
- **False Negatives**: Overly strict verifiers may filter working capabilities
- **Network Dependencies**: Verification requires external API access

### Medium Risk
- **Test Failures**: Updated verification behavior breaks existing tests
- **Credential Issues**: Invalid credentials cause capabilities to disappear
- **Memory Usage**: Health tracking adds memory overhead

### Low Risk
- **Backward Compatibility**: Changes are designed to be non-breaking
- **Performance**: Verification only runs at startup and during reconciliation
- **Security**: No new credential storage or exposure

## Rollback Plan

1. **Feature Flag**: Add environment variable to disable verification
2. **Graceful Degradation**: Fall back to old detection-only behavior
3. **Monitoring**: Track capability availability before/after deployment
4. **Quick Revert**: Revert to previous commit if critical issues arise

## Success Metrics

- **Reduced Job Failures**: Fewer jobs sent to workers with broken capabilities
- **Improved Network Stability**: More reliable capability advertising
- **Faster Problem Detection**: Broken capabilities identified at startup
- **Better Observability**: Real-time capability health visibility

## Next Steps

1. **Address Critical TODOs**: Fix missing verifiers and test issues
2. **Performance Testing**: Measure startup time impact
3. **Documentation**: Complete capability verification guide
4. **Deployment Planning**: Plan staged rollout strategy
5. **Monitoring Setup**: Prepare dashboards and alerts

### CODE REVIEW

File: internal/jobs/linkedin.go [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+func NewLinkedInScraper(jc types.JobConfiguration, c *stats.StatsCollector, h health.CapabilityHealthTracker) *LinkedInScraper {
+	config := types.LinkedInScraperConfiguration{}
+	if err := jc.Unmarshal(&config); err != nil {
+		logrus.Errorf("Error unmarshalling LinkedIn scraper configuration: %v", err)
+		return nil
+	}
+
+	// Parse credentials from environment variables if not in config
+	if len(config.Credentials) == 0 {
+		liAtCookie, _ := jc["linkedin_li_at_cookie"].(string)
```
Comment (rapidfix) [inline internal/jobs/linkedin.go:37]: suggestion: instead of navigating the job configuration as a map[]interface{}, use the JobConfiguration `UnMarshal` method to a concrete, defined typed struct: https://github.com/masa-finance/tee-worker/blob/6acf215b521cc2ae339101e95411cb924b699dbd/api/types/job.go#L113

File: internal/jobs/linkedin.go [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+		csrfToken, _ := jc["linkedin_csrf_token"].(string)
+		jsessionID, _ := jc["linkedin_jsessionid"].(string)
+
+		if liAtCookie != "" && csrfToken != "" {
+			config.Credentials = append(config.Credentials, types.LinkedInCredential{
```
Comment (rapidfix) [inline internal/jobs/linkedin.go:42]: I see you already have a types.LinkedinCredential. This can be directly used with the JobConfiguration Unmarshal

File: internal/capabilities/detector.go [hunk 3]
```diff
@@ -1,34 +1,214 @@
...
+		}
+
+		logrus.WithField("verifier", config.Name).WithField("capabilities", config.Capabilities).Debug("Registered verifier")
+	}
+
+	return verifiersMap
+}
+
+// hasLinkedInCredentials checks if LinkedIn credentials are available in any format
+func hasLinkedInCredentials(jc types.JobConfiguration) bool {
```
Comment (rapidfix) [inline internal/capabilities/detector.go:126]: ditto here, this naturally would belong to `types.LinkedInCredential`
